src/protocols/vnc/auth.c

```diff
@@ -19,27 +19,96 @@
     guac_client* gc = rfbClientGetClientData(client, GUAC_VNC_CLIENT_KEY);
-    return ((guac_vnc_client*) gc->data)->settings->password;
+    guac_vnc_client* vnc_client = ((guac_vnc_client*) gc->data);
+    guac_vnc_settings* settings = vnc_client->settings;
+    
+    /* If the client does not support the "required" instruction, just return
+        the configuration data. */
+    if (!guac_client_owner_supports_required(gc))
+        return ((guac_vnc_client*) gc->data)->settings->password;
+    
+    /* If password isn't around, prompt for it. */
+    if (settings->password == NULL || strcmp(settings->password, "") == 0) {
+        
+        guac_argv_register(GUAC_VNC_ARGV_PASSWORD, guac_vnc_argv_callback, NULL, 0);
+        
+        const char* params[] = {GUAC_VNC_ARGV_PASSWORD, NULL};
+        
+        /* Send the request for password to the owner. */
+        guac_client_owner_send_required(gc, params);
+                
+        /* Wait for the arguments to be returned */
+        guac_argv_await(params);
+
+    }
+    
+    return settings->password;
+    
 }
 
 rfbCredential* guac_vnc_get_credentials(rfbClient* client, int credentialType) {
     guac_client* gc = rfbClientGetClientData(client, GUAC_VNC_CLIENT_KEY);
-    guac_vnc_settings* settings = ((guac_vnc_client*) gc->data)->settings;
+    guac_vnc_client* vnc_client = ((guac_vnc_client*) gc->data);
+    guac_vnc_settings* settings = vnc_client->settings;
     
+    /* Handle request for Username/Password credentials */
     if (credentialType == rfbCredentialTypeUser) {
         rfbCredential *creds = malloc(sizeof(rfbCredential));
-        creds->userCredential.username = settings->username;
-        creds->userCredential.password = settings->password;
-        return creds;
+        
+        /* If the client does not support the "required" instruction, just return
+           the parameters already configured. */
+        if (!guac_client_owner_supports_required(gc)) {
+            creds->userCredential.username = settings->username;
+            creds->userCredential.password = settings->password;
+            return creds;
+        }
+        
+        char* params[2] = {NULL};
+        int i = 0;
+        
+        /* Check if username is null or empty. */
+        if (settings->username == NULL) {
+            guac_argv_register(GUAC_VNC_ARGV_USERNAME, guac_vnc_argv_callback, NULL, 0);
+            params[i] = GUAC_VNC_ARGV_USERNAME;
+            i++;
+        }
+        
+        /* Check if password is null or empty. */
+        if (settings->password == NULL) {
+            guac_argv_register(GUAC_VNC_ARGV_PASSWORD, guac_vnc_argv_callback, NULL, 0);
+            params[i] = GUAC_VNC_ARGV_PASSWORD;
+            i++;
+        }
+        
+        /* If we have empty parameters, request them. */
+        if (i > 0) {
+            
+            /* Send required parameters to owner. */
+            guac_client_owner_send_required(gc, (const char**) params);
```

Comment:
`guac_client_owner_send_required()` requires a null-terminated array. This will be the case here if only one parameter is requested, but because the length of `params` is 2, it cannot be the case if two parameters are requested, and the function may overrun array bounds.

I think your `params` array needs to be one element longer to ensure there is room for the NULL terminator.